### PR TITLE
Merge support for unknown item count to master

### PR DESCRIPTION
**Pull request by pleku**
Grid part in https://github.com/vaadin/vaadin-grid-flow/pull/1056

### CODE REVIEW

File: flow-data/src/main/java/com/vaadin/flow/data/provider/DataCommunicator.java [hunk 1]
```diff
@@ -393,16 +636,18 @@ protected Object getFilter() {
...
-                            getDataProvider().getClass());
+                    getDataProvider().getClass());
             stream = stream.collect(Collectors.toList()).stream();
             assert !stream.isParallel();
         }
+
         SizeVerifier verifier = new SizeVerifier<>(limit);
         stream = stream.peek(verifier);
 
+        // FIXME simplify by removing these restrictions ?
```
Comment (mshabarov) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/DataCommunicator.java:650]: Could we remove it now, or is it better to create a ticket to improve this in the future?

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasLazyDataView.java [hunk 2]
```diff
@@ -16,63 +16,112 @@
...
      */
-    V setDataProvider(
-            CallbackDataProvider.FetchCallback<T, Void> fetchCallback);
+    default V setDataSource(
+            CallbackDataProvider.FetchCallback<T, Void> fetchCallback) {
+        setDataSource(DataProvider.fromCallbacks(fetchCallback, query -> {
+            throw new IllegalStateException(
+                    "Trying to use defined size with a lazy loading component"
+                            + " without either providing a count callback for the"
+                            + "component to fetch the size of the data or a data"
```
Comment (mshabarov) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/HasLazyDataView.java:57]:
```suggestion
                            + " component to fetch the size of the data or a data"
```

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasLazyDataView.java [hunk 3]
```diff
@@ -16,63 +16,112 @@
...
+                            + "provider that implements the size query. Provide the "
```
Comment (mshabarov) [inline flow-data/src/main/java/com/vaadin/flow/data/provider/HasLazyDataView.java:58]:
```suggestion
                            + " provider that implements the size query. Provide the "
```

**Comment by mshabarov** [comment]
Probably the components submodules were not deleted completely, because i see the reference to grid repo and the component folders in the Flow project root dir.

File: flow-data/src/main/java/com/vaadin/flow/data/provider/DataCommunicator.java (hunk 1 above)
Comment (pleku) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/DataCommunicator.java]: Done

File: flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractDataView.java [hunk 4]
```diff
@@ -81,10 +81,13 @@ public Registration addSizeChangeListener(
...
      */
     protected final void verifyDataProviderType(Class<?> dataProviderType) {
+        // TODO https://github.com/vaadin/flow/issues/8583
         Class<?> supportedDataProviderType = getSupportedDataProviderType();
         if (!supportedDataProviderType.isAssignableFrom(dataProviderType)) {
             final String message = String.format(
-                    "%s only supports '%s' or it's subclasses, but was given a '%s'",
+                    "%s only supports '%s' or it's subclasses, but was given a '%s'."
+                            + "%nUse either 'getLazyDataView()', 'getListDataView()'"
+                            + " or getDataView() according to the used data type.",
```
[the remark this answers is not in the document]
Comment (pleku) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractDataView.java]: Done

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasLazyDataView.java (hunk 2 above)
Comment (pleku) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/HasLazyDataView.java]: Done

File: flow-data/src/main/java/com/vaadin/flow/data/provider/HasLazyDataView.java (hunk 3 above)
Comment (pleku) [reply flow-data/src/main/java/com/vaadin/flow/data/provider/HasLazyDataView.java]: Done

**Pull request merged**